File: cli/src/reporter.ts
import process from 'node:process';

type Verb =
  | 'Translating'
  | 'Revisiting'
  | 'Validating'
  | 'Checking'
  | 'Ok'
  | 'Stale'
  | 'Missing'
  | 'Removed'
  | 'Skipped'
  | 'Cleaned'
  | 'Created'
  | 'Updated'
  | 'Summary'
  | 'Info'
  | 'Dry run';

const COLORS = {
  reset: '\u001b[0m',
  bold: '\u001b[1m',
  dim: '\u001b[2m',
  green: '\u001b[32m',
  cyan: '\u001b[36m',
  yellow: '\u001b[33m',
  red: '\u001b[31m',
  white: '\u001b[37m',
} as const;

function colorForVerb(verb: Verb): string {
  switch (verb) {
    case 'Ok':
    case 'Removed':
    case 'Cleaned':
    case 'Created':
    case 'Updated':
      return COLORS.green;
    case 'Translating':
    case 'Revisiting':
    case 'Validating':
    case 'Checking':
      return COLORS.cyan;
    case 'Stale':
    case 'Skipped':
    case 'Dry run':
      return COLORS.yellow;
    case 'Missing':
      return COLORS.red;
    default:
      return COLORS.white;
  }
}

function formatVerb(verb: Verb, useColor: boolean): string {
  const padded = verb.padStart(12, ' ');
  if (!useColor) {
    return padded;
  }

  return `${COLORS.bold}${colorForVerb(verb)}${padded}${COLORS.reset}`;
}

export interface Reporter {
  log(verb: Verb, message: string): void;
  step(verb: Verb, current: number, total: number, message: string): void;
  blank(): void;
}

export class ConsoleReporter implements Reporter {
  readonly #useColor: boolean;

  constructor(noColor = false) {
    this.#useColor = !noColor && Boolean(process.stdout.isTTY);
  }

  log(verb: Verb, message: string): void {
    process.stdout.write(`${formatVerb(verb, this.#useColor)}  ${message}\n`);
  }

  step(verb: Verb, current: number, total: number, message: string): void {
    const stepLabel = `[${current}/${total}] ${message}`;
    this.log(verb, stepLabel);
  }

  blank(): void {
    process.stdout.write('\n');
  }
}

File: cli/src/reporter.rs
use std::fmt;
use std::io::{self, IsTerminal, Write};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Verb {
    Created,
}

impl fmt::Display for Verb {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let label = match self {
            Self::Created => "Created",
        };
        write!(f, "{label}")
    }
}

pub trait Reporter {
    fn log(&mut self, verb: Verb, message: &str);
    fn blank(&mut self);
}

pub struct ConsoleReporter {
    use_color: bool,
}

impl ConsoleReporter {
    pub fn new(no_color: bool) -> Self {
        Self {
            use_color: !no_color && io::stdout().is_terminal(),
        }
    }

    fn color_for_verb(&self, verb: Verb) -> &'static str {
        if !self.use_color {
            return "";
        }

        match verb {
            Verb::Created => "\u{001b}[1m\u{001b}[32m",
        }
    }

    fn format_verb(&self, verb: Verb) -> String {
        let padded = format!("{verb:>12}");
        if !self.use_color {
            return padded;
        }
        format!("{}{}\u{001b}[0m", self.color_for_verb(verb), padded)
    }
}

impl Reporter for ConsoleReporter {
    fn log(&mut self, verb: Verb, message: &str) {
        let mut stdout = io::stdout().lock();
        let _ = writeln!(stdout, "{}  {}", self.format_verb(verb), message);
    }

    fn blank(&mut self) {
        let mut stdout = io::stdout().lock();
        let _ = writeln!(stdout);
    }
}
